### PR TITLE
SDIT-1593: ✨ Add in mock user and context facade

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsAuthContextFacade.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package uk.gov.justice.hmpps.kotlin.auth
+
+import org.springframework.security.core.Authentication
+import org.springframework.security.core.GrantedAuthority
+import org.springframework.security.core.context.SecurityContextHolder
+import org.springframework.stereotype.Component
+
+@Component
+class HmppsAuthContextFacade {
```
Comment (petergphillips) [inline hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsAuthContextFacade.kt:9]: copied from Prison API

File: hmpps-kotlin-spring-boot-test-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/test/kotlin/auth/JwtAuthorisationHelper.kt [hunk 2]
```diff
@@ -45,7 +50,7 @@ class JwtAuthorisationHelper {
       scope = scope,
       roles = roles,
     )
-    return { it.set(HttpHeaders.AUTHORIZATION, "Bearer $token") }
+    return { it.setBearerAuth(token) }
```
Comment (petergphillips) [inline hmpps-kotlin-spring-boot-test-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/test/kotlin/auth/JwtAuthorisationHelper.kt:53]: found neater day of doing it

File: hmpps-kotlin-spring-boot-test-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/test/kotlin/auth/WithMockAuthUser.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+@Target(
+  AnnotationTarget.FUNCTION,
+  AnnotationTarget.PROPERTY_GETTER,
+  AnnotationTarget.PROPERTY_SETTER,
+  AnnotationTarget.CLASS,
+)
+@Retention(AnnotationRetention.RUNTIME)
+@Inherited
+@WithSecurityContext(factory = WithMockUserSecurityContextFactory::class)
+annotation class WithMockAuthUser(
```
Comment (petergphillips) [inline hmpps-kotlin-spring-boot-test-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/test/kotlin/auth/WithMockAuthUser.kt:28]: copied from Prison API

File: hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsAuthContextFacade.kt [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  val isClientOnly: Boolean
+    get() = authentication?.isSystemClientCredentials() ?: false
+
+  val clientId: String?
+    get() = authentication?.clientId
+
+  /**
+   * We are gradually moving away from user tokens and instead using client credentials more often.  This will be NONE
+   * if client credentials are used, even if a NOMIS username is passed in.  This means that this method isn't an
+   * adequate test to see if the user is a NOMIS user. The only test at present is to look the user up in the database
```
Comment (mikehalmamoj) [inline hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsAuthContextFacade.kt:37]: Looking up in the database won't be possible for most clients of the library - better to signpost to the nomis users API? (sorry can't remember which one that is!)

File: hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsAuthContextFacade.kt (hunk 1 above)
Comment (mikehalmamoj) [reply hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsAuthContextFacade.kt:9]: The name's a bit technical - I'm not sure I understand what a context facade actually is. Or maybe just needs a paragraph describing what the class is for?

File: hmpps-kotlin-spring-boot-test-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/test/kotlin/auth/WithMockAuthUser.kt [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import org.springframework.security.test.context.support.WithSecurityContext
+import org.springframework.security.test.context.support.WithSecurityContextFactory
+import org.springframework.util.Assert
+import uk.gov.justice.hmpps.kotlin.auth.AuthAwareAuthenticationToken
+import uk.gov.justice.hmpps.kotlin.auth.AuthSource
+import java.lang.annotation.Inherited
+
+/**
+ * Similar to {@link org.springframework.security.test.context.support.WithMockUser} but
+ * populates the security context with a {@link uk.gov.justice.hmpps.kotlin.auth.AuthAwareAuthenticationToken} instead.
```
Comment (mikehalmamoj) [inline hmpps-kotlin-spring-boot-test-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/test/kotlin/auth/WithMockAuthUser.kt:17]: Could we maybe expand a little on when/why you would use the annotation in a test?

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**